File: packages/mock-server/src/routes/settlements.ts
import { type Router as ExpressRouter, Router } from "express";
import { generateId } from "../lib/id-generator.js";
import { saveResource, getResource, listResources, updateResource } from "../lib/store.js";
import { emitEvent } from "../lib/event-bus.js";
import { withIdempotency } from "../middleware/idempotency.js";

const router: ExpressRouter = Router();

router.post("/", withIdempotency, (req, res) => {
  const now = new Date().toISOString();
  const settlement = saveResource("settlement", {
    id: generateId("stl"),
    object: "settlement",
    payee_id: req.body.payee_id,
    destination_id: req.body.destination_id ?? null,
    amount: req.body.amount,
    currency: req.body.currency,
    status: "draft",
    transaction_ids: req.body.transaction_ids ?? [],
    provenance: null,
    metadata: req.body.metadata ?? null,
    submitted_at: null,
    succeeded_at: null,
    created_at: now,
    updated_at: now,
    livemode: false,
  });
  emitEvent("settlement.created", "settlement", settlement.id, settlement);
  res.status(201).json(settlement);
});

router.get("/", (req, res) => {
  const result = listResources("settlement", {
    limit: req.query.limit ? Number(req.query.limit) : undefined,
    starting_after: req.query.starting_after as string | undefined,
  });
  res.json(result);
});

router.get("/:id", (req, res) => {
  const settlement = getResource("settlement", req.params.id);
  if (!settlement) return res.status(404).json(notFound("settlement", req.params.id));
  res.json(settlement);
});

router.post("/:id/submit", withIdempotency, (req, res) => {
  const settlement = getResource<{ status: string }>("settlement", req.params.id);
  if (!settlement) return res.status(404).json(notFound("settlement", req.params.id));

  if (settlement.status !== "draft") {
    return res.status(409).json({
      error: {
        code: "state.invalid_transition",
        http_status: 409,
        message: `Cannot submit a settlement with status '${settlement.status}'.`,
        retryable: false,
      },
    });
  }

  const now = new Date().toISOString();
  const updated = updateResource("settlement", req.params.id, {
    status: "submitted",
    submitted_at: now,
  });
  emitEvent("settlement.submitted", "settlement", req.params.id, updated);

  // Simulate async processing — immediately succeed in mock
  setTimeout(() => {
    const succeeded = updateResource("settlement", req.params.id, {
      status: "succeeded",
      succeeded_at: new Date().toISOString(),
    });
    emitEvent("settlement.succeeded", "settlement", req.params.id, succeeded);
  }, 100);

  res.json(updated);
});

router.post("/:id/cancel", withIdempotency, (req, res) => {
  const settlement = getResource<{ status: string }>("settlement", req.params.id);
  if (!settlement) return res.status(404).json(notFound("settlement", req.params.id));

  const cancelableStatuses = new Set(["draft"]);
  if (!cancelableStatuses.has(settlement.status)) {
    return res.status(409).json({
      error: {
        code: "state.invalid_transition",
        http_status: 409,
        message: `Cannot cancel a settlement with status '${settlement.status}'.`,
        retryable: false,
      },
    });
  }

  const updated = updateResource("settlement", req.params.id, { status: "canceled" });
  res.json(updated);
});

function notFound(resource: string, id: string) {
  return {
    error: { code: `${resource}.not_found`, http_status: 404, message: `${resource} ${id} not found.` },
  };
}

export default router;

File: packages/python-sdk/src/certifieddata_agent_commerce/resources/settlements.py
"""Settlements resource."""

from typing import Any, Optional
from .._http import HttpClient


class SettlementsResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def create(
        self,
        *,
        payee_id: str,
        amount: int,
        currency: str,
        transaction_ids: list[str],
        destination_id: Optional[str] = None,
        metadata: Optional[dict[str, str]] = None,
        idempotency_key: Optional[str] = None,
    ) -> dict[str, Any]:
        return self._http.request(
            "POST",
            "/v1/settlements",
            json={
                "payee_id": payee_id,
                "amount": amount,
                "currency": currency,
                "transaction_ids": transaction_ids,
                "destination_id": destination_id,
                "metadata": metadata,
            },
            idempotency_key=idempotency_key,
        )

    def get(self, settlement_id: str) -> dict[str, Any]:
        return self._http.request("GET", f"/v1/settlements/{settlement_id}")

    def list(
        self,
        *,
        limit: Optional[int] = None,
        starting_after: Optional[str] = None,
    ) -> dict[str, Any]:
        return self._http.request(
            "GET",
            "/v1/settlements",
            params={"limit": limit, "starting_after": starting_after},
        )

    def submit(
        self,
        settlement_id: str,
        *,
        idempotency_key: Optional[str] = None,
    ) -> dict[str, Any]:
        return self._http.request(
            "POST",
            f"/v1/settlements/{settlement_id}/submit",
            idempotency_key=idempotency_key,
        )

    def cancel(
        self,
        settlement_id: str,
        *,
        idempotency_key: Optional[str] = None,
    ) -> dict[str, Any]:
        return self._http.request(
            "POST",
            f"/v1/settlements/{settlement_id}/cancel",
            idempotency_key=idempotency_key,
        )
